source/elements/oneDNN/source/graph/programming_model.rst

```diff
@@ -0,0 +1,374 @@
+implementation dependent whether the compilation succeed. If the compilation
+succeeds for unknown dimension or rank, the compiled partition should be able to
+handle any value for that dimension or any rank at the execution time. 
+
+
+------
+Tensor
+------
+
+*Tensor* is an abstraction for multidimensional input and output data needed in
+the execution of a compiled partition. A tensor contains a logical tensor, an
+engine and a data handle.
+
+Users are responsible for managing the tensor’s lifecycle, e.g. free the
+resource allocated, when it is not used anymore.
+
+
+------------------
+Compiled Partition
+------------------
+
+A *compiled partition* represents the generated code specialized for target
+hardware and meta data described by parameter logical tensors. Compiled
+partition contains a partition and a handle representing the target specific
+compiled object.
+
+After the compilation API is invoked, users must query the logical output tensor
+of the compiled partition to know the output tensor’s layout id and size. The
+layout id is an opaque identifier for the target-specific layout. Users may pass
+the layout id for the next partition compilation so that it can be optimized to
+expect a specific input layout.  Users may use the size to allocate the memory
+buffer of the output tensors for execution.
+
+Framework passes the tensors and compiled partition as parameters to execution
+API. The parameter logical tensors must be in the same order when they are
+passed in the compilation API, and their IDs must match with the compiled
+partition’s internal logical tensors. The layout type of each tensor must be
+``strided`` or ``opaque``.
+
+The compiled partition may support in-place optimization, which reuses the input
+tensor data buffer for the output tensor for lower memory footprint and better
+data locality. For each compiled partition, users can get pairs of input and
+output ports. For the pair of input and output ports, user can use a same memory
+buffer when passing input and output tensors along with execution API. The
+in-place optimization is optional, when users use another memory buffer for the
+output tensor, oneDNN Graph must update the output tensor.
+
+If users place a tensor with data buffer pointer in outputs, the backend shall
+use the data buffer provided by users.
+
+Users may convert the parameter tensor with public layout to the target specific
+layout expected by the compiled partition. A common optimization in deep
+learning inference is that users may prepack the weight in the target-specific
+layout required by the compiled partition and cache the reordered weight for
+late use.
+
+------
+Engine
+------
+
+*Engine* (|engine|) are an abstraction of a computational device. The
+ graph extension additionally allows to create an engine with specific
+ host/device allocators to conveniently manage memory inside the Graph
+ API calls.
```

Comment:
There are extra spaces in the start of lines which will cause rendering issue.

```suggestion
*Engine* (|engine|) are an abstraction of a computational device. The
graph extension additionally allows to create an engine with specific
host/device allocators to conveniently manage memory inside the Graph
API calls.
```

![image](https://user-images.githubusercontent.com/34727741/200988108-f3bf0139-ad60-4853-a00c-01e07d7674c8.png)

Reply:
Thanks for the catch!